PMNS_SNSI.cxx

```diff
@@ -0,0 +1,165 @@
+////////////////////////////////////////////////////////////////////////
+//
+// Implementation of oscillations of neutrinos with scalar NSI
+// three-neutrino framework. 
+//
+// jcoelho@apc.in2p3.fr and urahaman@km3net.de
+////////////////////////////////////////////////////////////////////////
+
+#include "PMNS_SNSI.h"
+
+using namespace OscProb;
+
+//......................................................................
+///
+/// Constructor. \sa PMNS_NSI::PMNS_NSI
+///
+/// This class is restricted to 3 neutrino flavours.
+///
+PMNS_SNSI::PMNS_SNSI() : PMNS_NSI()
+{
+  SetLowestMass(0);
+}
+
+//......................................................................
+///
+/// Nothing to clean.
+///
+PMNS_SNSI::~PMNS_SNSI(){}
+
+//......................................................................
+///
+/// Set lightest neutrino mass
+///
+/// @param m - The lightest mass in eV
+///
+void PMNS_SNSI::SetLowestMass(double m)
+{
+  fM = m;
+}
+
+//......................................................................
+///
+/// Get lightest neutrino mass
+///
+/// @return The lightest mass in eV
+///
+double PMNS_SNSI::GetLowestMass()
+{
+  return fM;
+}
+
+//......................................................................
+///
+/// Reimplement vacuum Hms update
+///
+/// \sa PMNS_Base::BuildHms
+///
+void PMNS_SNSI::BuildHms()
+{
+
+  // Check if anything changed
+  if(fBuiltHms) return;
+  
+  // Tag to recompute eigensystem
+  fGotES = false;
+ 
+  // Start with m1 = 0
+  double m1 = 0;
+  // Make sure all masses are positive
+  for(int i=1; i<fNumNus; i++){
+    if(fDm[i] + m1*m1 < 0) m1 = sqrt(fabs(fDm[i]));
+  }
+  // Add minimum mass
+  m1 += fM;
+
+  // Build M - m1
+  fHms[0][0] = 0;
+  for(int j=1; j<fNumNus; j++){
+    // Set mass difference m_j - m_1
+    fHms[j][j] = sqrt(fDm[j] + m1*m1) - m1;
+    // Reset off-diagonal elements
+    for(int i=0; i<j; i++){
+      fHms[i][j] = 0;
+    }
+    // Rotate j neutrinos
+    for(int i=0; i<j; i++){
+      RotateH(i,j,fHms);
+    }
+  }
+ 
+  // Add back m1
+  for(int i=0; i<fNumNus; i++){
+    fHms[i][i] += m1;
+  }
+
+  ClearCache();
```

Comment:
I suspect the caching of eigensystems may not work with this class, because we are sensitive to more than just 2EVe. Maybe we need to disable caching.